Add Credentials SPI Extension

vsevel opened the pull request:
Extracts the `CredentialsProvider` interface and move it into its own `credentials` extension, to prepare for more provider implementations (either custom, or provided by quarkus as other extensions, such as support for Azure Vault).

In this PR we will also 

* support the interface in the reactive pg client, as discussed in https://github.com/quarkusio/quarkus/issues/8826#issuecomment-619927442
* document how to implement a new provider (custom or qurkus based)

Fixes https://github.com/quarkusio/quarkus/issues/8828 and https://github.com/quarkusio/quarkus/issues/8826.

see also 

 * [datasource credentials from Azure KeyVault](https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/datasource.20credentials.20from.20Azure.20KeyVault), 
 * [custom provider example](https://github.com/gsmet/quarkus-credentials-provider)
 * https://github.com/quarkusio/quarkus/issues/6896

vsevel committed:
Add Credentials SPI Extension

gsmet commented on integration-tests/pom.xml:
```diff
@@ -62,6 +62,7 @@
         <module>oidc-tenancy</module>
         <module>keycloak-authorization</module>
         <module>reactive-pg-client</module>
+        <module>reactive-pg-client-creds</module>
```
Let's use `credentials` instead of `creds` please.

vsevel replied:
the test has been moved to the deployment module as suggested later on

gsmet reviewed (changes requested):
Looks mostly good. Added one small comment. Also, we should do the same work for the Reactive MySQL client? It's exactly the same one as the PostgreSQL client.

What bugs me a bit is we add 10mn of native executable testing for a very narrow feature. But I don't see how we could test both normal password setup and credential setup with the same IT. @geoand any idea?

Adding multiple datasources support for the reactive clients would definitely help :).

geoand commented:
> What bugs me a bit is we add 10mn of native executable testing for a very narrow feature. But I don't see how we could test both normal password setup and credential setup with the same IT. @geoand any idea?

Not really. If it were jus the JVM tests we could probably do stuff, but with the native test I don't see any good way to do it.

gsmet commented:
I don't think there's anything specific to native so maybe we could put the tests in the deployment dir for now?

And then we add native tests once we have the multiple datasource support for the reactive clients?

geoand commented:
Makes sense to add a set of native tests when the feature set becomes available, instead of just adding a tiny one now.

I agree that the deployment tests should be fine for now.

geoand commented:
Don't forget to squash the commits :wink:

vsevel commented:
hello @gsmet and @geoand and thanks for your reviews.
your comments about the shape and locations of the tests were completely appropriate. I went through difficulties when doing the development of the test initially, especially when trying to run it on macos for some reason (I will come back on this later). And since I wanted a build to run after the initial dev, I ended up duplicating the existing IT test, and removing the WIP from the issue.
however, there are a few things I would like to ask you before considering this for a merge.

first I was surprised that I had to mark the `CustomCredentialsProvider` as `@Unremovable` or `@Named`, otherwise it would not be found by the `PgPoolRecorder`. I would have though that `@ApplicationScoped` would be enough.
is this expected?

second, I am surprised that `skip=false` currently on module `reactive-pg-client/deployment`. as a result tests do not execute even with flags `-Dtest-postgresql -Ddocker`. see https://github.com/quarkusio/quarkus/commit/9ff7545e695c454d646c3db09729c9a896ab4207
have I missed something? should not that be `skip=true` when we activate `-Dtest-postgresql `. 
as a result this [build](https://pipelines.actions.githubusercontent.com/5Na9LjmLgXLiLuzJ0Z1qHRbIDVUXUHNrKe6jPcqGdw5t1S8TiW/_apis/pipelines/1/runs/24457/signedlogcontent/21?urlExpires=2020-05-05T06%3A32%3A57.4306740Z&urlSigningMethod=HMACV1&urlSignature=ymANs9SFl6YDSAxY4qHl93Pq6eso07Sk54PFgiUkbFM%3D) shows that tests have been skipped:

```
2020-05-04T21:56:42.4017084Z [INFO] --- maven-surefire-plugin:2.22.1:test (default-test) @ quarkus-reactive-pg-client-deployment ---
2020-05-04T21:56:42.4033582Z [INFO] Tests are skipped.
``` 

on my todo list, I have noted:

- [ ] look at the `reactive-mysql-client`
- [ ] provide some documentation around building a custom provider
- [ ] simplify the unit test I have added to the deployment module (right now it is roughly a copy/paste of the `FruitsEndpointTest`)

FYI I also changed the design a little bit where `getCredentials()` is directly called from the recorder, instead of from a supplier set into the pg options. The reason is that the credentials provider might return a user and a password. so we need to make one call, and configure both on the options. specifically, we cannot make 2 calls to get the user then the password, because the user and the password have to go side by side, and might be regenerated (such as in vault's dynamic db credentials). this means that the user/password will be retrieved only once for the process. I do not see any other way, without making changes to the vertx code. 

lastly I am getting this error when running my test on macos (and not on ubuntu):

```
[INFO] Running io.quarkus.reactive.pg.client.FruitsEndpointCredentialsTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 5.778 s <<< FAILURE! - in io.quarkus.reactive.pg.client.FruitsEndpointCredentialsTest
[ERROR] io.quarkus.reactive.pg.client.FruitsEndpointCredentialsTest  Time elapsed: 5.775 s  <<< ERROR!
java.lang.RuntimeException: java.lang.IllegalArgumentException: Undeclared build item class io.quarkus.deployment.builditem.ApplicationClassNameBuildItem
        at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:438)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$7(ClassBasedTestDescriptor.java:359)
...
Caused by: java.lang.IllegalArgumentException: Undeclared build item class io.quarkus.deployment.builditem.ApplicationClassNameBuildItem
        at io.quarkus.builder.Messages.undeclaredItem(Messages.java:13)
        at io.quarkus.builder.BuildResult.consume(BuildResult.java:48)
        at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:157)
        at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:398)
        ... 37 more
```

does it ring any bell?

once I have addressed all the points, I will do the final squash. thanks for the reminder.

gsmet commented:
So I'm not entirely sure your -deployment module for credentials is included in the dependencies. I don't know exactly how the runtime module is linked to the deployment one but I'm not sure we can have a `-spi` suffix. @aloubyansky should the runtime dependency always be named `<dependency>` for `<dependency>-deployment` to be taken into account?

gsmet commented:
Also I was wondering why we are returning `Properties` instead of a bean? Do we expect more information to be added? Even so, I think I would favor a bean to which we could add a hash map if needed.

Or maybe we wouldn't provide username and password in some cases?

vsevel commented:
> but I'm not sure we can have a -spi suffix

this was a straight renaming from the original `quarkus-vault-spi`. I do not know if there are any constraints on `runtime` modules that would make them not fitted to be used from user's code when implementing a custom provider. a `runtime` module sounds good to be if we can.

> why we are returning Properties instead of a bean

this comes from agroal accepting a `Properties` object. it could be a `Map<String, String>`

> Or maybe we wouldn't provide username and password in some cases?

it is really dependent on the consumer. some will accept a password, and some will accept a token of some sort, or any other mean to authenticate. so to make it evolutive, I did not think we could set the fields in stone. the compromise was to add key constants for common use cases (ie user and password).

all of these are valid questions.

aloubyansky commented:
The "spi" artifact is not going to be recognized as a Quarkus extension. See https://quarkus.io/guides/writing-extensions#maven-setup how to configure the runtime artifact.
Now, this could be an interesting case. I guess there could be technical extensions that don't have any runtime artifact. I can see that a bunch of extensions depend on the `credentials-deployment`, which will make it a part of the build, so I guess it could be ok.
But I don't think we've had any extension like that to this day.

aloubyansky commented:
As to the naming, assuming it's going to be a true runtime artifact of an extension, it could be called anything. But we do strongly encourage the naming convention documented in the guide at this point.

geoand commented:
Seems like CI is complaining

vsevel commented:
> Seems like CI is complaining

there was no filtering on the test resources. not sure we really should be doing this into the deployment module itself. I guess the problem was already there but not seen, snice the tests were skipped.

vsevel commented:
new CI failure because postgresql is not running for JVM tests. should I move the test back to the integration tests (without making it native)?

geoand commented:
Yeah, that sounds more reasonable at this point than trying to hack around everything

gsmet commented:
> new CI failure because postgresql is not running for JVM tests. should I move the test back to the integration tests (without making it native)?

Yeah those reactive tests are really bad. We should make them depend of a running cluster and have them executed only with a `test-postgresql` profile similar to the `jpa-postgresql` tests.

That's a lot of cleanup unrelated to your main task though. Do you want me to work on that part in another PR and have it merged so that you can focus on your task?

> runtime module

So I would drop the `-spi` suffix for now and make it a standard extension. That will be safer if people want to use it directly

> `Properties`

Considering we are making it a general purpose extension, I think we should have a clean API and only use properties in Agroal.

vsevel commented:
ok @geoand and @gsmet I will implement all of your different suggestions.
then we will be able to do another round of reviews.

vsevel commented:
@geoand I was trying to add the new test into existing module `quarkus-integration-test-reactive-pg-client`
since I have a new application.properties (to add the credentials provider properties) different from the existing one currently in `src/main/properties`, I started moving the existing one as 'application-base.properties' and adding my new one in `src/test/properties`, then dropped the `@QuarkusTest`, and replaced with in existing tests:

```
    @RegisterExtension
    static final QuarkusUnitTest config = new QuarkusUnitTest()
            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
                    .addAsResource("application-base.properties", "application.properties"));
```

however I get an error on existing tests:

```
INFO] Running io.quarkus.it.reactive.pg.client.FruitsEndpointTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.04 s <<< FAILURE! - in io.quarkus.it.reactive.pg.client.FruitsEndpointTest
[ERROR] io.quarkus.it.reactive.pg.client.FruitsEndpointTest  Time elapsed: 1.04 s  <<< ERROR!
java.lang.ExceptionInInitializerError
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:398)
...
aused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:169)
        ... 41 more
Caused by: java.lang.RuntimeException: RESTEASY003130: Class is not a root resource.  It, or one of its interfaces must be annotated with @Path: io.quarkus.it.reactive.pg.client.FruitResource implements: 
        at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:206)
        at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:192)
        at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:175)
        at org.jboss.resteasy.core.ResourceMethodRegistry.addPerRequestResource(ResourceMethodRegistry.java:87)
        at org.jboss.resteasy.core.ResteasyDeploymentImpl.registerResources(ResteasyDeploymentImpl.java:513)
        at org.jboss.resteasy.core.ResteasyDeploymentImpl.registration(ResteasyDeploymentImpl.java:470)
        at org.jboss.resteasy.core.ResteasyDeploymentImpl.startInternal(ResteasyDeploymentImpl.java:160)
        at org.jboss.resteasy.core.ResteasyDeploymentImpl.start(ResteasyDeploymentImpl.java:117)
        at io.quarkus.resteasy.runtime.standalone.ResteasyStandaloneRecorder.staticInit(ResteasyStandaloneRecorder.java:86)
        at io.quarkus.deployment.steps.ResteasyStandaloneBuildStep$staticInit-210558872.deploy_0(ResteasyStandaloneBuildStep$staticInit-210558872.zig:863)
        at io.quarkus.deployment.steps.ResteasyStandaloneBuildStep$staticInit-210558872.deploy(ResteasyStandaloneBuildStep$staticInit-210558872.zig:36)
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:152)
        ... 41 more
```

is there a way to make it work? thanks for your help.

gsmet commented:
No really, what we need is to have them here: https://github.com/quarkusio/quarkus/tree/master/extensions/reactive-pg-client/deployment/src/test/java/io/quarkus/reactive/pg/client .

I'll work on fixing the reactive tests tomorrow morning so that you can add yours there.

I'll ping you in the PR.

geoand commented:
Don't forget to squash the commits 😉

vsevel commented:
@gsmet https://github.com/quarkusio/quarkus/pull/9119 does work on my local tests. waiting for the CI to confirm.
I will be able to continue from there on the other items discussed previously. thanks.

vsevel commented:
a quick update:

> So I would drop the -spi suffix for now and make it a standard extension. That will be safer if people want to use it directly

done

> Considering we are making it a general purpose extension, I think we should have a clean API and only use properties in Agroal.

I replaced the `Properties` by a `Map<String, String>`. do you want to go one step further and have a bean instead with fields password and user, which is sufficient today, but might prove too restricted when new use cases pop up?

> No really, what we need is to have them here: https://github.com/quarkusio/quarkus/tree/master/extensions/reactive-pg-client/deployment/src/test/java/io/quarkus/reactive/pg/client .

done

> msql

done

next:
 - decide on Map<String, String> vs Credentials bean
 - doc
 - centralize logic to get access to credentials provider

sberyozkin commented:
Hi @vsevel, all

> I replaced the Properties by a Map<String, String>. do you want to go one step further and have a bean instead with fields password and user, which is sufficient today, but might prove too restricted when new use cases pop up?

IMHO `Map` is good. If we introduce the typed getPassword etc then it would be not great if some credentials providers uses some other form of credentials, so this bean would still need to have `getMap` to accommodate such cases.

Vincent, can you please resolve a conflict

vsevel commented:
hi @sberyozkin 
thanks for your review
I will work on doc this week-end

gsmet commented:
I force pushed to fix a conflict. Let's wait for CI and merge.

vsevel commented:
> I force pushed to fix a conflict. Let's wait for CI and merge.

@gsmet don't you want me to provide a guide, or at least some javadoc on the  `CredentialsProvider`?

gsmet commented:
@vsevel some doc would be welcome but we can merge it a bit later. I release CR1 on Tuesday so I want the code in before that. Then let's get the doc before end of month.

gsmet commented:
Merged, thanks! I let you work on the appropriate documentation before the Final release.
